db/attachment.go

```diff
@@ -125,6 +130,53 @@ func (db *Database) storeAttachments(doc *Document, newAttachmentsMeta Attachmen
 	return newAttachmentData, nil
 }
 
+// retrieveObsoleteAttachments determines the list of attachments that are no longer being referenced from any document
+// and returns a slice of attachments keys for subsequent garbage collection.
+func (db *Database) retrieveObsoleteAttachments(doc *Document, docAttachments AttachmentsMeta, parentRev string, docHistory []string) (obsoleteAttachments []string, err error) {
```

Comment:
This is comparing attachments on the current rev with attachments on the current rev's ancestor.  I don't think this accounts for attachments associated with conflicting revisions (which will be stored as _attachment in the conflicting revision body, see storeOldBodyInRevTreeAndUpdateCurrent).  I think we want to compare the set (current doc attachments + all conflicting rev body attachments) with (new doc attachments + all conflicting rev body attachments).  Otherwise we risk deleting an attachment referenced on a non-winning revision, and not have a way to recreate that attachment if that revision becomes the active rev.

Similarly, I don't think parentRev should be used, as that's not necessarily the parent of the active revision (e.g. the case where someone updates a conflicting branch, but the new revision is still on the losing branch).   There are quite a few permutations there that should get unit test coverage.

Reply:
Obsolete attachment retrieval logic has been refactored and the retrieveObsoleteAttachments can be completely removed. The retrieveActiveAttachments has been remaned to retrieveAttachments and included the versions check while marking obsolete attachments.